prow/git/git_command_test.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"reflect"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/util/diff"
+)
+
+func urlMustParse(u string) *url.URL {
+	ret, err := url.Parse(u)
+	if err != nil {
+		panic(err)
+	}
+	return ret
+}
+
+type tmpdirMock struct {
+	basedir   string
+	readerIdx int
+}
+
+func newTmpdirMock() *tmpdirMock {
+	basedir, err := ioutil.TempDir("", "mock")
+	if err != nil {
+		panic(err)
+	}
+	return &tmpdirMock{basedir: basedir}
+}
+
+func (t *tmpdirMock) recorded(idx int) string {
+	return fmt.Sprintf("%s/%d", t.basedir, idx)
+}
+
+func (t *tmpdirMock) TempDir(dir, prefix string) (ret string, err error) {
+	ret = t.recorded(t.readerIdx)
+	t.readerIdx++
+	return
+}
+
+func TestClone(t *testing.T) {
+	gitbin, err := exec.LookPath("git")
+	if err != nil {
+		t.Errorf("Git is not found: %v", err)
+	}
+
+	tempdir := newTmpdirMock()
+	ioutilTempDir = tempdir.TempDir
+	defer func() { ioutilTempDir = ioutil.TempDir }()
+
+	cases := []struct {
+		name           string        //name of the test
+		base           string        //Client parameters
+		user           string        //if user is not empty, SetCredentials will be called
+		tokenGenerator func() []byte //credentials token generator func
+
+		repo                 string     // what repo do we clone?
+		simulateExecFailures int        // how many exec failures?
+		commands             [][]string // what commands does this run?
+		expected             Repo       // what repo does it return?
+		err                  bool       // did it return any error?
+	}{
+		{
+			name:                 "happy path with remote url and user",
+			base:                 "https://github.com",
+			user:                 "user",
+			tokenGenerator:       func() []byte { return []byte("password") },
+			repo:                 "org/repo",
+			simulateExecFailures: 0,
```

Comment:
This is a poor way to make things fail as it is very fragile to the implementation of the method under test and hard to keep straight for someone adding or changing functionality. You have to remember what order commands are called and how many retries each has. You also cannot make an intermediate command fail after previous commands succeeded.